### PR TITLE
python script enhancements - fix missing Element ID when order

### PR DESCRIPTION
**Pull request by bonidier**
Hi @bittner,

Following remaining part in issue #8, here is the pull request to map missing element ID with a chain of updated ID ... and more !

Regards,
Didier

### CODE REVIEW

File: lego-mindstorms-pieces.py [hunk 1]
```diff
@@ -90,6 +90,28 @@ def main():
     function(**kwargs)
 
 
+def load_new_element_ids():
+    """
+    load list of new element IDs
+    """
+    global newelementid_map
+    global newelementid_comment
```
Comment (bittner) [inline lego-mindstorms-pieces.py]: I cannot allow code with `global` to slip in. ([Why are global variables evil?](https://stackoverflow.com/questions/19158339/why-are-global-variables-evil#19158418))

Please create a class if you need to share values between functions.

File: lego-mindstorms-pieces.py [hunk 2]
```diff
@@ -339,67 +365,132 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
...
 
-        element_field.clear()
-        element_field.send_keys(part_no)
-        element_field.send_keys(Keys.RETURN)
-        sleep(.3)  # seconds
+        # never add the same part twice,
+        # otherwise the quantity will be set to the previous part
+        if part_no in added_part.keys():
+            duplicate_part_counter += 1
+            print("IGNORE: Already added !".format(pn=part_no))
```
Comment (bittner) [inline lego-mindstorms-pieces.py]: French spacing again ! (Got you!) :smile:

File: lego-mindstorms-pieces.py [hunk 3]
```diff
@@ -339,67 +365,132 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
...
+                        break
+
+                    # we're on the original part, and we've a list of new Element ID
+                    if idx == 0 and len(partno_list) > 1:
+                        print("Not Found, but has a chain of other Element ID:")
+                        # a comment about the mapping
+                        if part_no in newelementid_comment.keys():
+                            print("\tcomment: {comment}".
+                                  format(comment=newelementid_comment[original_part_no]))
+
```
Comment (bittner) [inline lego-mindstorms-pieces.py]: This empty line seems unnecessary.

File: lego-mindstorms-pieces.py [hunk 4]
```diff
@@ -339,67 +365,132 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
...
+                    print("Found!")
+
+                    add_button = browser.find_element_by_css_selector('.element-details + button')
+
+                    if add_button.is_enabled():
+                        add_button.click()
+                        sleep(.2)  # seconds
+                        found_counter += 1
+
+                        # quit the part_no chain loop
```
Comment (bittner) [inline lego-mindstorms-pieces.py]: This comment is a (relatively strong) indication that the loop body is significantly too long. Try moving some parts of the loop body into helper functions to make the code easier to understand. When everything is logical, even without comments, then you're good.

**Review by bittner — CHANGES_REQUESTED** [review]

**Comment by bonidier** [comment]
Hi @bittner,

To reduce the loop, I'd moved some code outside, but not in a helper : there is too many variables to manage.

It's more readable for now.

The order() function is already long. In my mind it can be better (in a near future) to refactor the content under OOP practice, for maintainability (like using a method per step)

File: lego-mindstorms-pieces.py (hunk 2 above)
Comment (bonidier) [reply lego-mindstorms-pieces.py]: Man, "you really got me" :musical_note:

**Comment by bittner** [comment]
Wow! The `order` function is now _huge_! Can you split this up a bit?

I think such a huge function reduces readability and thus both maintainability and the probability of future contributions. We need to take care this code is in good shape. Minimum split long function bodies up into several functions, or even move large pieces that belong together in separate classes and/or modules.

I don't want to be the dog in the manger, but I only have two choices: Either oppose and continue maintaining this repo, or let this code creep into the repo and let maintenance go.  :cry:

**Comment by bonidier** [comment]
Hi @bittner,
According to our common idea, I'll split the order() function for maintenance purpose

**Comment by bonidier** [comment]
Hi @bittner,
Here is a first try to move order() content in a module

Regards !

**Comment by bittner** [comment]
Not bad! Starts to look nice, doesn't it? -- (**NOTE:** _It will be more complicated for Python beginners, because there is more than a single script to download, but that's the price for getting more features._)

Try to move the helper functions you have (the ones that start with `_` and `__`) out into a helper module or so. And don't forget to run [flake8](http://flake8.pycqa.org/en/latest/index.html#quickstart) over your code! Gives you great feedback.

**Comment by bonidier** [comment]
Hi Peter, 
I don't forget your feedback about helper module suggestion.
Here is some fixes about Lego's website evolutions since last commit

**Comment by bittner** [comment]
Thanks for your follow-up. Please, give me a few days for reviewing.

Thank you for your understanding.

**Comment by bonidier** [comment]
Hi @bittner, 
Thanks for your reactivity, take your time.

Just a note about survey: Lego has changed their survey form, from now we must play with HTML area tag,
I don't now how to fix it, my non-working try is commented.

fortunately, the survey does not appears at each execution

**Review by bittner — APPROVED** [review]

**Comment by bittner** [comment]
I'm sorry I obviously forgot to merge these changes! Should I merge them now?

**Comment by bonidier** [comment]
Hi, thanks for your approval!
I've just relaunched order for Home/Education sets for a last review:
Here is a fix for set 45544 about missing ElementIDs to add manually (out of the set).

Now, you can merge
Regards!

**Comment by bittner** [comment]
Sorry for the delay. I didn't get the GitHub notifications for this repo for some reason. Strange.

**Comment by bittner** [comment]
Thanks for your contribution! :+1:

**Pull request merged**